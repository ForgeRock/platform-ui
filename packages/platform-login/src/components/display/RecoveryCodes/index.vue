<!--
Copyright (c) 2020 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details.
-->
<template>
  <div>
    <h2>
      {{ $t('login.recoveryCodes.header') }}
    </h2>
    <p class="mb-0">
      {{ $t('login.recoveryCodes.subheader') }}
    </p>
    <BRow>
      <BCol
        cols="4"
        offset="4">
        <hr class="my-4">
      </BCol>
    </BRow>
    <h5 class="text-left">
      {{ $t('login.recoveryCodes.subtitle') }}
    </h5>
    <p class="text-left">
      {{ $t('login.recoveryCodes.info') }}
      <strong>
        {{ $t('login.recoveryCodes.infoEmphasis') }}
      </strong>
    </p>
    <div class="border rounded bg-light mb-4">
      <div class="p-4 text-left">
        <BRow>
          <BCol>
            <ol
              start="1"
              class="m-0 column-wrapper">
              <li
                v-for="(code, key) in recoveryCodes"
                :key="key">
                <span
                  class="ml-2 text-monospace">
                  {{ code }}
                </span>
              </li>
            </ol>
          </BCol>
        </BRow>
      </div>
    </div>
    <div
      class="d-flex flex-row-reverse justify-content-between">
      <BButton
        variant="primary"
        @click="$emit('next-step')"
      >
        {{ $t('common.done') }}
      </BButton>
      <div>
        <BButton
          variant="link"
          @click="copyRecoveryCodesToClipboard">
          {{ $t('common.copy') }}
        </BButton>
        <BButton
          variant="link"
          @click="printRecoveryCodes">
          {{ $t('common.print') }}
        </BButton>
      </div>
    </div>
  </div>
</template>

<script>
import { FRRecoveryCodes } from '@forgerock/javascript-sdk';
import NotificationMixin from '@forgerock/platform-shared/src/mixins/NotificationMixin';
import { writeText } from 'clipboard-polyfill/text';
import {
  BCol,
  BRow,
  BButton,
} from 'bootstrap-vue';

export default {
  name: 'RecoveryComponent',
  mixins: [
    NotificationMixin,
  ],
  components: {
    BCol,
    BRow,
    BButton,
  },
  props: {
    step: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      recoveryCodes: [],
    };
  },
  mounted() {
    this.recoveryCodes = FRRecoveryCodes.getCodes(this.step);
  },
  methods: {
    copyRecoveryCodesToClipboard() {
      writeText(this.recoveryCodes).then(() => {
        this.displayNotification('', 'success', this.$t('common.copySuccess'));
      }, (error) => {
        this.showErrorMessage(error, this.$t('common.copyFail'));
      });
    },
    printRecoveryCodes() {
      window.print();
    },
  },
};
</script>

<style lang="scss" scoped>
.column-wrapper {
  column-count: 2;

  ::marker {
    color: $gray-400;
  }
}
</style>
