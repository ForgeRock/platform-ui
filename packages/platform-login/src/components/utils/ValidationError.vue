<template>
  <div
    v-show="validatorErrors.length"
    class="fr-validation-requirements text-left">
    <p
      v-for="(error) in validatorErrors"
      :key="error"
      class="text-danger mb-0">
      {{ error }}
    </p>
  </div>
</template>

<script>
// TODO Add to components
/**
 * @description Component for displaying error message for form fields
 *
 * */
export default {
  name: 'ValidationError',
  props: {
    validatorErrors: {
      type: Array,
      default: () => [],
    },
    fieldName: {
      type: String,
      default: '',
    },
  },
};
</script>

<style lang="scss" scoped></style>
