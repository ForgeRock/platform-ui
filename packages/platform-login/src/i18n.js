/**
 * Copyright (c) 2020-2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Vue from 'vue';
import VueI18n from 'vue-i18n';
import { merge } from 'lodash';
import dateTimeFormats from '@forgerock/platform-shared/src/dateTimeFormats.json';

Vue.use(VueI18n);

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  // Can't use workspace symlink due to jest issues
  const sharedLocales = require.context('../../platform-shared/src/locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      let generatedTranslation = locales(key);

      // Check for shared local and merge in if it exists
      const sharedLocaleExists = sharedLocales.keys().includes(key);
      if (sharedLocaleExists) {
        generatedTranslation = merge(sharedLocales(key), generatedTranslation);
      }

      messages[locale] = generatedTranslation;
    }
  });

  return messages;
}

Vue.mixin({
  methods: {
    translationExists(path) {
      return this.$t(path) !== path;
    },
  },
});

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages(),
  silentTranslationWarn: true,
  dateTimeFormats,
});
