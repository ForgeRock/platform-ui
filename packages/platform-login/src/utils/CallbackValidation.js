import { isArray } from 'lodash';

/**
 * @description Util "class" specifically for validating callback output and input
 *
 * */
export default {
	name: 'CallbackValidation',
	// make sure the callback has an output or input property and the input or output is an Array that has at least one item
	validateOutput: prop => prop.output
            && isArray(prop.output)
            && prop.output.length > 0,
	validateInput: prop => prop.input
            && isArray(prop.input)
            && prop.input.length > 0,
};
