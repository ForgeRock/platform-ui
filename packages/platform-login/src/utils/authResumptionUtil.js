/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

const reentryTokenName = 'reentry';
const resumeDataKey = 'treeResumeData';
const authResumptionParams = ['state', 'code', 'scope', 'form_post_entry', 'responsekey'];

/**
 * Returns boolean true if reentry cookie is set
 * @returns {Boolean}
 */
function hasReentryToken() {
  return !!document.cookie
    .split('; ')
    .find((row) => row.startsWith(reentryTokenName));
}

/**
 * Clears the reentry token by updating the cookie to expire in the past
 */
function clearReentryToken() {
  const date = new Date();
  date.setTime(date.getTime() + (-1 * 24 * 60 * 60 * 1000));
  document.cookie = `reentry="";expires="${date.toGMTString()}";path=/`;
}

/**
 * Determines whether the passed url parameters contain any of the queries indicating a tree is being resumed
 * @param {URLSearchParams} urlParams the query parameters for the current login session
 * @returns {Boolean} whether the passed urlParams include any resumption query parameters
 */
function doURLParamsContainAnyResumptionParameter(urlParams) {
  const paramKeyArray = Array.from(urlParams.keys());
  return paramKeyArray.some((queryParam) => authResumptionParams.includes(queryParam));
}

/**
 * Checks whether any resumption data has been stored
 * @returns {Boolean} whether any resumption data is present in browser storage
 */
function isResumeDataPresentInStorage() {
  return localStorage.getItem(resumeDataKey) !== null;
}

/**
 * Determines whether the passed url parameters and browser storage indicate that a tree is being resumed following a redirect
 * @param {URLSearchParams} urlParams the query parameters for the current login session
 * @returns {Boolean} whether a tree should be resumed
 */
function resumingTreeFollowingRedirect(urlParams) {
  // The AM XUI just checks for a reentry token and data in browser storage here. The check against query parameters should not be needed but are left for extra security
  return (doURLParamsContainAnyResumptionParameter(urlParams) || hasReentryToken()) && isResumeDataPresentInStorage();
}

/**
 * Determines whether the passed route name and url parameters indicate that we're resuming a suspended tree
 * @param {String} routeName the name of the current route
 * @param {URLSearchParams} urlParams the query parameters for the current login session
 */
function resumingSuspendedTree(routeName, urlParams) {
  return routeName === 'login' && urlParams.has('suspendedId') && urlParams.has('authIndexValue');
}

/**
 * Adds data for resuming a tree into browser storage for later retrieval to resume a tree
 * @param {Step} step the SDK step to store
 */
function addTreeResumeDataToStorage(step, realmAtRedirect) {
  const storeData = {
    realmAtRedirect,
    step,
  };
  localStorage.setItem(resumeDataKey, JSON.stringify(storeData));
}

/**
 * Loads tree resumption data from browser storage and removes it, also clearing the tracking cookie
 * @returns {Object} an object containing the url at redirect, and the sdk step for resuming a tree
 */
function getResumeDataFromStorageAndClear() {
  const resumeData = localStorage.getItem(resumeDataKey);
  localStorage.removeItem(resumeDataKey);
  clearReentryToken();

  return JSON.parse(resumeData);
}

export {
  hasReentryToken,
  clearReentryToken,
  doURLParamsContainAnyResumptionParameter,
  isResumeDataPresentInStorage,
  resumingTreeFollowingRedirect,
  resumingSuspendedTree,
  addTreeResumeDataToStorage,
  getResumeDataFromStorageAndClear,
};
