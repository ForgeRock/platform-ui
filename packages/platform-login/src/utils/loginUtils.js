/**
 * Copyright (c) 2023-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

/**
 * Contains the list of form labels and their corresponding `autocomplete` value.
 * https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete
 */
const autocompleteFieldLookup = {
  // identification
  username: 'username',
  'user name': 'username',
  password: 'current-password',
  prefix: 'honorific-prefix',
  'first name': 'given-name',
  'last name': 'family-name',
  'middle name': 'additional-name',
  'display name': 'nickname',
  nickname: 'nickname',
  'email address': 'email',
  'street address': 'street-address',
  // Phone
  'telephone number': 'tel',
  'phone number': 'tel',
  'country code': 'tel-country-code',
  'area code': 'tel-area-code',
  'phone extension': 'tel-extension',
  // preferences
  language: 'language',
  'preferred language': 'language',
};

/**
 * Finds the autocomplete value from the given label.
 * For accessibility purposes certain fields should have an `autocomplete` attribute.
 * @param {String} fieldName the label of the form field
 * @returns autocomplete value for that label
 */
function getAutocompleteValue(fieldName) {
  if (!fieldName) return false;
  if (typeof fieldName !== 'string') return false;

  return autocompleteFieldLookup[fieldName.toLowerCase()];
}

export default getAutocompleteValue;
