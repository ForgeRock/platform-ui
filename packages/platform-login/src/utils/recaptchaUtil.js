/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

/**
 * Assesses whether the passed node list contains elements created to display a google recaptcha v2 challenge
 * @param {NodeList} addedNodeList the new nodes viewed by a MutationObserver
 * @returns {Boolean} whether the new nodes are part of a google recaptcha v2 node
 */
function doNewNodesContainRecaptchaV2(addedNodeList) {
  try {
    if (addedNodeList.length !== 1) return false;
    const iframes = addedNodeList[0].querySelectorAll('iframe');
    if (iframes.length !== 1) return false;
    return iframes[0].src.includes('www.google.com/recaptcha');
  } catch {
    return false;
  }
}

export default doNewNodesContainRecaptchaV2;
