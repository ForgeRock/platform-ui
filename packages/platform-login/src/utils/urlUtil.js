/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import {
  fromPairs, isEmpty, map,
} from 'lodash';

/**
 * Returns the query string from the URI
 * @returns {string} Unescaped query string or empty string if no query string was found
 */
function getCurrentQueryString() {
  const queryString = window.location.search;
  return queryString.substring(1, queryString.length);
}

/**
 * @description Creates an object of key value pairs from the passed in query string
 * @param {string} paramString A string containing a query string
 * @returns {object} An Object of key value pairs
 */
function parseParameters(paramString) {
  const object = isEmpty(paramString) ? {} : fromPairs(map(paramString.split('&'), (pair) => pair.split('=')));
  return object;
}

/**
 * Creates a query string from the passed urlParams, starting with & rather than ?
 * @param {URLSearchParams} urlParams to create a query string from)
 * @returns {String} a query string
 */
function createParamString(urlParams) {
  let stringParams = '';

  urlParams.forEach((value, key) => {
    if (key === 'authIndexValue' && urlParams.get('authIndexType') === 'service') {
      // Don't encode the tree names in the URL as these won't be matched correctly if they are
      stringParams += `&${key}=${value}`;
    } else {
      stringParams += `&${encodeURIComponent(key)}=${encodeURIComponent(value)}`;
    }
  });

  return stringParams;
}

/**
 * Updates browser history so that the URL only contains the passed realm, tree and general parameters
 * @param {String} realm the realm the include in the URL
 * @param {URLSearchParams} newUrlParams the url params to include in the URL
 */
function replaceUrlParams(realm, newUrlParams) {
  const { location: { hash } } = window;
  const newParamString = createParamString(newUrlParams);
  window.history.replaceState(null, null, `?realm=${realm}${newParamString}${hash}`);
}

export {
  getCurrentQueryString,
  parseParameters,
  createParamString,
  replaceUrlParams,
};
