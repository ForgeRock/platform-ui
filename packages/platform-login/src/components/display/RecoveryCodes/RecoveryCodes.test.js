/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { mount } from '@vue/test-utils';
import { findByTestId } from '@forgerock/platform-shared/src/utils/testHelpers';
import { FRRecoveryCodes } from '@forgerock/javascript-sdk';
import * as clipboard from 'clipboard-polyfill/text';
import Notifications from '@kyvg/vue3-notification';
import i18n from '@/i18n';
import RecoveryCodes from './index';

describe('RecoveryCodes', () => {
  window.print = jest.fn();

  const stubRecoveryCodes = [
    'ZJrCaPOrP1',
    'GgyOtUBObK',
    'd6UFLvg7J1',
    'lW53KEMwZN',
    'Rzz77kf0B7',
    'Y2pcoQgMD9',
    '8bSsPn7AW6',
    'aglbEnRO0O',
    'j0Cdp7aoWQ',
    '8YrBjjy5Hl',
  ];

  jest.spyOn(FRRecoveryCodes, 'getCodes').mockReturnValue(stubRecoveryCodes);

  function setup(props) {
    return mount(RecoveryCodes, {
      global: {
        plugins: [i18n, Notifications],
      },
      props: {
        step: {},
        ...props,
      },
    });
  }

  describe('@renders', () => {
    it('recovery codes', async () => {
      const wrapper = setup();
      await wrapper.vm.$nextTick();

      stubRecoveryCodes.forEach((code, index) => {
        const li = findByTestId(wrapper, `recovery-code-${index}`);
        expect(li.text(code)).toBe(code);
      });
    });
  });

  describe('@actions', () => {
    it('can progress to next step', () => {
      const wrapper = setup();

      expect(wrapper.emitted('next-step')).toBeFalsy();
      const doneButton = findByTestId(wrapper, 'btn-recovery-codes-next-step');
      doneButton.trigger('click');

      expect(wrapper.emitted('next-step')).toBeTruthy();
    });

    it('should copy codes to clipboard', async () => {
      const wrapper = setup();
      await wrapper.vm.$nextTick();
      const notificationSpy = jest.spyOn(wrapper.vm, 'displayNotification');
      const clipboardSpy = jest.spyOn(clipboard, 'writeText').mockImplementation(() => Promise.resolve());

      const copyCodesButton = findByTestId(wrapper, 'btn-copy-recovery-codes');
      expect(copyCodesButton.attributes('aria-label')).toBe(`Copy ${stubRecoveryCodes.length} recovery codes to clipboard`);
      expect(notificationSpy).not.toHaveBeenCalled();
      expect(clipboardSpy).not.toHaveBeenCalled();

      await copyCodesButton.trigger('click');
      expect(notificationSpy).toHaveBeenCalledTimes(1);
      expect(clipboardSpy).toHaveBeenCalledWith(stubRecoveryCodes);
    });

    it('should display error when copy codes fails', async () => {
      jest.spyOn(clipboard, 'writeText').mockReturnValue(Promise.reject());

      const wrapper = setup();
      await wrapper.vm.$nextTick();
      const showErrorMessageSpy = jest.spyOn(wrapper.vm, 'showErrorMessage');

      const copyCodesButton = findByTestId(wrapper, 'btn-copy-recovery-codes');
      expect(showErrorMessageSpy).not.toHaveBeenCalled();

      await copyCodesButton.trigger('click');
      expect(showErrorMessageSpy).toHaveBeenCalledTimes(1);
    });

    it('should print codes', async () => {
      const print = jest.spyOn(window, 'print');

      const wrapper = setup();
      await wrapper.vm.$nextTick();

      const printCodesButton = findByTestId(wrapper, 'btn-print-recovery-codes');
      expect(printCodesButton.attributes('aria-label')).toBe(`Print ${stubRecoveryCodes.length} recovery codes`);
      expect(print).not.toHaveBeenCalled();

      await printCodesButton.trigger('click');
      expect(print).toHaveBeenCalledTimes(1);
    });
  });
});
