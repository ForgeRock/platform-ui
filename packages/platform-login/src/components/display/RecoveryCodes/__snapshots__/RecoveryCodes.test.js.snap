// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecoveryCodes @renders recovery codes: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) Functional component <BRow> should be defined as a plain function in Vue 3. The "functional" option has been removed. NOTE: Before migrating to use plain functions for functional components, first make sure that all async components usage have been migrated and its compat behavior has been disabled.
  Details: https://v3-migration.vuejs.org/breaking-changes/functional-components.html",
    "
",
    " at <RecoveryComponent",
    "step=",
    {},
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (2)",
    "
",
    " at <RecoveryComponent",
    "step=",
    {},
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) vm.$listeners has been removed. In Vue 3, parent v-on listeners are included in vm.$attrs and it is no longer necessary to separately use v-on="$listeners" if you are already using v-bind="$attrs". (Note: the Vue 3 behavior only applies if this compat config is disabled)
  Details: https://v3-migration.vuejs.org/breaking-changes/listeners-removed.html",
    "
",
    " at <BButton",
    "variant="primary"",
    "onClick=fn",
    "data-testid="btn-recovery-codes-next-step"",
    ">",
    "
",
    " at <RecoveryComponent",
    "step=",
    {},
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
